### PR TITLE
feat: SNDR CI fix, policy_train deprecation, Recommendation/DiagnosticGate APIs, coverage simulation harness (#58 #60 #62 #81 #82 #83 #99)

### PR DESCRIPTION
**Pull request by dgenio**
## Summary

Implements all 7 open issues in a single PR: SNDR CI correctness fix, `policy_train` deprecation, `Recommendation` and `DiagnosticGate` APIs, and a coverage-probability simulation harness.

---

### #58 — Fix SNDR bootstrap CI formula

**Bug**: The SNDR bootstrap pseudo-values were computed using the DR (unnormalised) formula `q_pi + w*(Y - q_hat)`. SNDR uses self-normalised importance weights, so the correct formula is:

```
pseudo = q_pi + (n * w_clip / w_sum) * (Y - q_hat)
```

**Fix**: `core.py` now uses `_sndr_bootstrap_values()` (shared helper) which branches on `estimator_name == "SNDR"` and applies the normalised formula in both `evaluate_sklearn_models` and `evaluate_pairwise_models`. Includes `w_sum == 0` guard.

**Validation**: `test_sndr_ci_covers_oracle_at_nominal_rate` proves SNDR CI brackets oracle V*_SNDR at ≥70% rate across B=50 seeds. `test_sndr_ci_covers_oracle_after_fix` confirms V_hat ∈ CI for all 30 seeds (consistency check).

---

### #60 — `policy_train` sentinel parameter

Both `evaluate_sklearn_models` and `evaluate_pairwise_models` now accept `policy_train: str | None = None`. Passing `None` (the legacy default) emits a `DeprecationWarning` and resolves to `"pre_split"`. Tests verify warn/no-warn paths.

---

### #62 — Coverage-probability simulation harness

New private module `src/skdr_eval/_simulation.py` with:
- `CoverageResult` dataclass
- `simulate_coverage(dgp, n, n_reps, alpha, ...)` — runs Monte-Carlo coverage experiments over three DGP families: `"iid"`, `"ar1"` (rho=0.5), `"seasonal"` (period=52)
- `Makefile` `coverage-sim` target

---

### #81 — Bootstrap CI coverage tests (DGP parametrized)

`TestBootstrapCICoverageDGP` in `test_bootstrap_integration.py`:
- DR CI coverage ≥70% across B=50 seeds (`test_dr_ci_covers_oracle_at_nominal_rate`)
- SNDR CI coverage ≥70% across B=50 seeds (`test_sndr_ci_covers_oracle_at_nominal_rate`)
- SNDR CI consistency: V_hat ∈ CI for B=30 seeds (`test_sndr_ci_covers_oracle_after_fix`)

---

### #82 — `policy_train` DeprecationWarning plumbing

Implementation detail of #60: the sentinel resolves before any validation. Existing tests that omit `policy_train` continue to work (they emit a warning). No `-W error::DeprecationWarning` in pytest config.

---

### #83 — `Recommendation` API

New public API in `reporting.py`:
- `RecommendationPolicy` — frozen dataclass holding `baseline` threshold
- `Reason` — dataclass with `code`, `message`, `severity` fields (structured message object, not an enum)
- `Recommendation` — dataclass with `verdict`, `confidence`, `primary_blocker`, `reasons`, `to_dict()`, `from_dict()`
- `EvaluationArtifact.recommendation(model_name, *, estimator, baseline, policy)` — method using `None` sentinel for `baseline` (allows explicit `baseline=0.0`)
- `_build_recommendation()` — internal decision logic

---

### #99 — `DiagnosticGate` API

- `GateResult` — per-metric gate outcome dataclass with `to_dict()`
- `DiagnosticGate` — aggregator with `overlap`, `ess`, `calibration`, `overall`, `to_dict()`, `to_text()`
- `gate_diagnostics(artifact, model_name, estimator, *, thresholds)` — convenience function

All new symbols exported from `__init__.py`.

---

## Test results

| Suite | Result |
|---|---|
| Full suite (Python 3.14) | **495 passed, 3 skipped** |
| CI (Python 3.11, 3.12, 3.13, 3.14) | **All green** |
| Ruff lint + format | **All checks passed** |

## Files changed

**Source**:
- `src/skdr_eval/core.py` — SNDR CI fix + `_sndr_bootstrap_values()` helper + `policy_train` sentinel
- `src/skdr_eval/reporting.py` — `Recommendation`, `DiagnosticGate` + `gate_diagnostics`
- `src/skdr_eval/_simulation.py` (**new**) — `simulate_coverage` harness
- `src/skdr_eval/__init__.py` — new exports

**Tests**:
- `tests/test_api.py` — `policy_train` warning tests + new symbol imports
- `tests/test_bootstrap_integration.py` — `TestBootstrapCICoverageDGP` (DR + SNDR coverage)
- `tests/test_coverage_simulation.py` (**new**) — coverage harness tests
- `tests/test_reporting_artifact.py` — `TestRecommendation` + `TestDiagnosticGate`
- `tests/test_temporal_split_controls.py` — `policy_train="all"` fix

**Docs / infra**:
- `CHANGELOG.md` — `[Unreleased]` section
- `Makefile` — `coverage-sim` target

### CODE REVIEW

File: src/skdr_eval/reporting.py [hunk 1]
```diff
@@ -1014,6 +1014,54 @@ def save_card(
...
+
+        Raises
+        ------
+        DataValidationError
+            If ``model_name`` or ``estimator`` is not found in the artifact.
+        """
+        _policy = RecommendationPolicy(baseline=baseline)
+        if policy is not None:
+            _policy = RecommendationPolicy(baseline=baseline if baseline != 0.0 else policy.baseline)
+        return _build_recommendation(self, model_name, estimator, _policy)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/reporting.py]: Fixed in 88cb0f7. Changed `baseline` parameter to `float | None = None` with proper precedence: when `baseline` is explicitly provided (including 0.0), it takes precedence; when `None`, falls back to `policy.baseline` if a policy is given, otherwise defaults to 0.0.

File: tests/test_reporting_artifact.py [hunk 2]
```diff
@@ -835,3 +837,359 @@ def test_card_handles_missing_diagnostics():
...
+
+
+from skdr_eval.reporting import (  # noqa: E402
+    DiagnosticGate,
+    GateResult,
+    Reason,
+    Recommendation,
+    RecommendationPolicy,
+    gate_diagnostics,
+)
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_reporting_artifact.py:854]: Fixed in 88cb0f7. Added `assert isinstance(gate, DiagnosticGate)` in `test_gate_pass_healthy_artifact` — resolves F401 and improves type coverage of the test.

File: src/skdr_eval/_simulation.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+          i.i.d. noise.
+    n : int, default 5000
+        Observations per replication.
+    n_reps : int, default 500
+        Number of Monte-Carlo replications.
+    alpha : float, default 0.05
+        Nominal significance level.  The target coverage is ``1 - alpha``.
+    block_length_strategy : {"auto", "fixed"}, default "auto"
+        ``"auto"`` lets :func:`block_bootstrap_ci` choose the block length
+        (currently ``max(1, n^{1/3})``).  ``"fixed"`` uses ``block_len``.
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/_simulation.py]: Fixed in 88cb0f7. Updated docstring to say `max(1, int(sqrt(n)))` matching the actual `block_bootstrap_ci` implementation in `core.py:1204`.

File: src/skdr_eval/reporting.py [hunk 4]
```diff
@@ -1236,6 +1284,483 @@ def _build_interpretation(
...
+    rows = artifact.report[row_mask]
+    if rows.empty:
+        raise DataValidationError(
+            f"No report row for model={model_name!r}, estimator={estimator!r}.",
+        )
+    row = rows.iloc[0]
+    support_health = str(row.get("support_health", SUPPORT_OK))
+    warn_str = str(row.get("diagnostic_warnings", "") or "")
+    warn_codes = [c for c in warn_str.split(",") if c]
+    v_hat = float(row["V_hat"])
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/reporting.py]: Fixed in 88cb0f7. Removed both unused assignments (`support_health` and `v_hat`). The downstream logic only references `warn_codes`, `ci_lower`, and `ci_upper`.

File: src/skdr_eval/_simulation.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    For each of ``n_reps`` replications, draws ``n`` observations from the
+    selected DGP, calls :func:`~skdr_eval.core.block_bootstrap_ci`, and
+    checks whether the true mean falls inside the resulting interval.
+
+    Parameters
+    ----------
+    dgp : {"iid", "ar1", "seasonal"}, default "ar1"
+        Data-generating process.
+
+        - ``"iid"``: independent standard-normal observations.
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/_simulation.py]: Fixed in 88cb0f7. Changed to "independent normal observations with mean=1.0, std=1.0" — aligns with `_sample_iid` which draws `N(1, 1)` so `true_mean=1.0`.

File: src/skdr_eval/core.py [hunk 6]
```diff
@@ -2040,6 +2088,20 @@ def evaluate_pairwise_models(
...
+        warnings.warn(
+            "evaluate_pairwise_models: policy_train was not specified and now "
+            "defaults to 'pre_split' (was 'all'). The 'all' mode fits and "
+            "evaluates on the same data, introducing training-on-test bias "
+            "when fit_models=True. Pass policy_train='pre_split' to suppress "
+            "this warning, or policy_train='all' to retain the old behavior. "
+            "See #60 and #82.",
+            DeprecationWarning,
+            stacklevel=2,
+        )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/core.py:2116]: Intentional — keeping the unconditional warning. The deprecation is about migrating all callers to an explicit `policy_train` value before it becomes required in a future release. Conditioning on `fit_models=True` would leave `fit_models=False` callers un-migrated, causing breakage when the parameter becomes required. The warning text already clarifies that the bias concern is specific to `fit_models=True`.

File: src/skdr_eval/_simulation.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    ----------
+    empirical_coverage : float
+        Fraction of replications in which the true parameter fell inside the
+        ``1 - alpha`` bootstrap CI.
+    ci_for_coverage : tuple[float, float]
+        Wilson (1927) 95% confidence interval for ``empirical_coverage``.
+    passes_nominal : bool
+        True when ``ci_for_coverage[0] >= alpha_tolerance``, i.e. the
+        empirical coverage is consistent with the nominal rate within the
+        sampling variability of the simulation.
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/_simulation.py]: Fixed in 88cb0f7. Updated `passes_nominal` docstring to describe the actual implementation: `empirical_coverage >= (1 - alpha) - tolerance`. The Wilson CI is computed and exposed but not used for gating (too conservative at small `n_reps`).

File: src/skdr_eval/reporting.py [hunk 8]
```diff
@@ -1236,6 +1284,483 @@ def _build_interpretation(
...
+    baseline: float = 0.0
+
+
+@dataclass
+class Reason:
+    """A single human-readable reason contributing to a :class:`Recommendation`.
+
+    Attributes
+    ----------
+    code : str
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/skdr_eval/reporting.py:1317]: Acknowledged — `Reason` is a structured dataclass, not an enum. The PR description will be refreshed in Phase 4 of this review cycle to accurately describe the implementation.

File: tests/test_coverage_simulation.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            warnings.warn(
+                f"Expected under-coverage for AR(1) with block_len=1 but got "
+                f"empirical={result.empirical_coverage:.2%}. "
+                "This may be a fluke at this seed; re-run with n_reps=500 to confirm.",
+                stacklevel=2,
+            )
+        # What we do assert: empirical coverage is at most the nominal + 5%
+        # (it should be noticeably below, but we bound from above to keep the
+        # test non-trivially informative).
+        assert result.empirical_coverage <= 0.95 + 0.05
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_coverage_simulation.py]: Fixed in 88cb0f7. Replaced with `assert not result.passes_nominal` — with the fixed seed (seed=0) and tolerance=0.02, this deterministically verifies that block_len=1 on AR(1) produces under-coverage. Removed the soft-warning fallback since the fixed seed makes the result deterministic.

File: tests/test_bootstrap_integration.py [hunk 10]
```diff
@@ -468,3 +470,128 @@ def test_pipeline_bootstrap_ci_sanity(self):
...
+            v = row["V_hat"]
+            lo = row["ci_lower"]
+            hi = row["ci_upper"]
+            if not (lo <= v <= hi):
+                bad += 1
+
+        assert bad == 0, (
+            f"SNDR CI failed to contain V_hat in {bad}/{B} seeds after issue #58 fix. "
+            "The SNDR CI pseudo-outcome may still be using DR contributions."
+        )
```
Comment (dgenio) [inline tests/test_bootstrap_integration.py:597]: **should-fix · tests/coverage gap**

The SNDR CI fix (#58) is a change to statistical evaluation logic. Per repo rules (`copilot-instructions.md` §1): "Any change to statistical evaluation logic MUST include a simulation that proves the code recovers a known ground-truth parameter."

`test_sndr_ci_covers_oracle_after_fix` only checks that `V_hat ∈ CI` for all 30 seeds — this is a **consistency** check (the point estimate should trivially be inside its own CI), not a **coverage-probability** simulation that proves the CI brackets the true population parameter at the nominal rate.

**Suggestion**: Add a companion test analogous to `test_dr_ci_covers_oracle_at_nominal_rate` but for SNDR: compute an SNDR-specific oracle V* and assert coverage ≥ 70% across B seeds. This would fully satisfy the repo's statistical integrity rule.

File: src/skdr_eval/core.py [hunk 11]
```diff
@@ -1489,10 +1519,24 @@ def evaluate_sklearn_models(
...
+                            w_sum = float(w_clip.sum())
+                            _n = len(eval_design.Y)
+                            if w_sum > 0:
+                                bootstrap_values = q_pi + (_n * w_clip / w_sum) * (
+                                    eval_design.Y - q_hat
+                                )
+                            else:
+                                bootstrap_values = q_pi.copy()
+                        else:
+                            bootstrap_values = q_pi + w_clip * (eval_design.Y - q_hat)
```
Comment (dgenio) [inline src/skdr_eval/core.py]: **minor · DRY / future-drift risk**

The SNDR pseudo-outcome computation (lines 1527–1536 here) is duplicated verbatim in `evaluate_pairwise_models` (lines ~2413–2435). If one copy is updated later (e.g., adding regularisation or a different normalisation), the other may be forgotten.

Consider extracting a shared helper:
```python
def _sndr_bootstrap_values(q_pi, w_clip, Y, q_hat):
    w_sum = float(w_clip.sum())
    n = len(Y)
    if w_sum > 0:
        return q_pi + (n * w_clip / w_sum) * (Y - q_hat)
    return q_pi.copy()
```

Non-blocking — the current code is correct. This is a maintainability suggestion for follow-up.

File: src/skdr_eval/_simulation.py [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+
+    # Wilson confidence interval for the coverage proportion
+    z = 1.959964  # z_{0.975}
+    n_r = n_reps
+    p = empirical_coverage
+    denom = 1.0 + z**2 / n_r
+    centre = (p + z**2 / (2 * n_r)) / denom
+    half = (z / denom) * math.sqrt(p * (1 - p) / n_r + z**2 / (4 * n_r**2))
+    ci_coverage = (max(0.0, centre - half), min(1.0, centre + half))
+
```
Comment (dgenio) [inline src/skdr_eval/_simulation.py:220]: **nit · hardcoded n_boot**

`n_boot=400` is hardcoded inside the simulation loop. For the CLI/programmatic API, consider exposing this as a parameter of `simulate_coverage()` — callers who want faster runs (CI) or more precise calibration (local) could tune it independently of `n_reps`.

Non-blocking — current default is reasonable for the intended use cases.

**Review by dgenio — COMMENTED** [review]
## Audit Review — PR #100

### Severity Summary

| Severity | Count | Files |
|----------|-------|-------|
| Blocker | 0 | — |
| Major | 0 | — |
| Should-fix | 2 | tests/test_bootstrap_integration.py, src/skdr_eval/core.py |
| Nit | 1 | src/skdr_eval/_simulation.py |

### Vibe-Coded Detection Results

| Signal | Status | Details |
|--------|--------|---------|
| V1 Missing tests | PASS | All new logic (Recommendation, DiagnosticGate, coverage sim) has dedicated test suites |
| V2 Happy-path-only | PASS | Tests cover error paths, edge cases, round-trip serialization |
| V3 Spec drift | PASS | All 7 issues implemented as described; no hidden TODOs |
| V4 Optimistic dependencies | PASS | Proper try/except fallback in bootstrap CI computation |
| V5 Swallowed errors | PASS | Errors propagated; DeprecationWarning uses stacklevel=2 |
| V6 Boundary bugs | PASS | w_sum=0 guard prevents div-by-zero; off-by-one review clean |
| V7 N+1 / hot-path waste | NA | No DB/network calls in this PR |
| V8 Overengineering | PASS | API surface matches stated requirements |
| V9 Security gap | NA | No user-facing input surfaces |
| V10 Cross-file inconsistency | PASS | __init__.py exports match new symbols; test imports align |
| V11 Goal not achieved | PASS | All 7 issues addressed; SNDR fix, policy_train, Recommendation, DiagnosticGate, simulation |
| V12 Suboptimal strategy | PASS | Design choices are sound for the domain |

### Summary

**Goal Achievement & Strategy**
- All 7 linked issues (#58 #60 #62 #81 #82 #83 #99) are implemented and tested
- The SNDR CI fix uses the correct self-normalised pseudo-outcome formula with proper w_sum=0 guard
- The `policy_train` deprecation follows Python's standard sentinel+warning pattern

**Correctness & Bugs**
- SNDR pseudo-outcome formula `q_pi + (n/Σw)*w*(Y - q_hat)` is mathematically correct for the self-normalised DR estimator
- Wilson CI for coverage proportion is properly computed
- No correctness issues found

**Architecture & Design**
- Clean separation: `_simulation.py` (private module), `reporting.py` (public APIs), `core.py` (fix in-place)
- DiagnosticGate and Recommendation compose well with existing EvaluationArtifact
- One DRY concern: SNDR bootstrap code duplicated across both evaluators (see inline comment)

**Tests, Performance & Robustness**
- 495 tests pass locally (3 skipped). CI GREEN after lint fixes.
- One statistical gap: SNDR CI coverage-probability test missing (see inline comment) — the existing test proves consistency (V_hat ∈ CI) but not nominal-rate coverage. This is a should-fix per repo rules.

### Decision

**Would APPROVE** — No blockers or major issues. The SNDR coverage test gap is a should-fix that strengthens statistical confidence but doesn't block merge given (a) the DR coverage test exists at the same code path, (b) the V_hat∈CI check validates the formula correction, and (c) the generic `simulate_coverage` harness validates the bootstrap mechanism independently.

(Submitted as COMMENT because GitHub does not allow self-approval.)

File: src/skdr_eval/core.py (hunk 11 above)
Comment (dgenio) [reply src/skdr_eval/core.py]: Fixed in 8c03f78. Extracted `_sndr_bootstrap_values(q_pi, w_clip, Y, q_hat)` helper near `block_bootstrap_ci` in core.py. Both `evaluate_sklearn_models` and `evaluate_pairwise_models` now call the shared helper.

File: src/skdr_eval/_simulation.py (hunk 12 above)
Comment (dgenio) [reply src/skdr_eval/_simulation.py:220]: Acknowledged — intentionally leaving as-is for now. The 400 bootstrap samples is a reasonable default for this harness's purpose (CI verification, not production inference). Can be parametrised in a follow-up if needed.

File: tests/test_bootstrap_integration.py (hunk 10 above)
Comment (dgenio) [reply tests/test_bootstrap_integration.py:597]: Fixed in 8c03f78. Added `test_sndr_ci_covers_oracle_at_nominal_rate` — computes SNDR oracle V* on full data (policy_train='all') and asserts the SNDR CI from pre_split evaluation brackets it at ≥70% rate across B=50 seeds. This satisfies the statistical integrity rule.
